lib/workers/repository/update/branch/execute-pre-upgrade-commands.ts

```diff
@@ -0,0 +1,193 @@
+import is from '@sindresorhus/is';
+import minimatch from 'minimatch';
+import { mergeChildConfig } from '../../../../config';
+import { GlobalConfig } from '../../../../config/global';
+import { addMeta, logger } from '../../../../logger';
+import type { ArtifactError } from '../../../../modules/manager/types';
+import { exec } from '../../../../util/exec';
+import {
+  localPathIsFile,
+  readLocalFile,
+  writeLocalFile,
+} from '../../../../util/fs';
+import { getRepoStatus } from '../../../../util/git';
+import type { FileChange } from '../../../../util/git/types';
+import { regEx } from '../../../../util/regex';
+import { sanitize } from '../../../../util/sanitize';
+import { compile } from '../../../../util/template';
+import type { BranchConfig, BranchUpgradeConfig } from '../../../types';
+
+export interface PreUpgradeCommandsExecutionResult {
+  updatedArtifacts: FileChange[];
+  artifactErrors: ArtifactError[];
+}
+
+export async function preUpgradeCommandsExecutor(
```

Comment:
refactor post upgrade tasks to be more generic, so we don't need to duplicate all that code for pre upgrade tasks.

refactor should probably be done in a separate PR.

Reply:
I have an another branch with refactoring (https://github.com/damien-danglard/renovate/pull/2). I can merge it in this one or make an another PR, as you want.

I think it's better to make an another PR because it's a big refactoring.

Reply:
yes, create a separate PR for refactoring first.
so we don't have to duplicate so much code

Reply:
I do this : https://github.com/renovatebot/renovate/pull/20571
@viceice  can you review it please ?